cassiopeia/data/CassiopeiaTree.py

```diff
@@ -0,0 +1,712 @@
+    """An Exception class for the CassiopeiaTree class."""
+
+    pass
+
+
+class CassiopeiaTree:
+    """Basic tree object for Cassiopeia.
+
+    This object stores the key attributes and functionalities a user might want
+    for working with lineage tracing experiments. At its core, it stores
+    three main items - a tree, a character matrix, and meta data associated
+    with the data.
+
+    The tree can be fed into the object via Ete3, Networkx, or can be inferred
+    using one of the CassiopeiaSolver algorithms in the `solver` module.
+
+    A character matrix can be stored in the object, containing the states
+    observed for each cell. In typical lineage tracing experiments, these are
+    integer representations of the indels observed at each unique cut site.
+
+    Meta data for cells or characters can also be stored in this object. These
+    items can be categorical or numerical in nature. Common examples of cell
+    meta data are the cluster identity, tissue identity, or number of target-site
+    UMIs per cell. These items can be used in downstream analyses, for example
+    the FitchCount algorithm which infers the number of transitions between
+    categorical variables (e.g., tissues). Common examples of character meta
+    data are the proportion of missing data for each character or the entropy
+    of states. These are good statistics to have for feature selection.
+
+    TODO(mattjones315): Add experimental meta data as arguments.
+    TODO(mattjones315): Add functionality that mutates the underlying tree
+        structure: pruning lineages, collapsing mutationless edges, and
+        collapsing unifurcations
+    TODO(mattjones315): Add utility methods to compute the colless index
+        and the cophenetic correlation wrt to some cell meta item
+
+    Args:
+        character_matrix: The character matrix for the lineage.
+        missing_state_indicator: An indicator for missing states in the
+            character matrix.
+        cell_meta: Per-cell meta data
+        character_meta: Per-character meta data
+        priors: A dictionary storing the probability of a character mutating
+            to a particular state.
+        tree: A tree for the lineage. 
+    """
+
+    def __init__(
+        self,
+        character_matrix: Optional[pd.DataFrame] = None,
+        missing_state_indicator: int = -1,
+        cell_meta: Optional[pd.DataFrame] = None,
+        character_meta: Optional[pd.DataFrame] = None,
+        priors: Optional[Dict[int, Dict[int, float]]] = None,
+        tree: Optional[Union[str, ete3.Tree, nx.DiGraph]] = None,
+    ):
+
+        self.__original_character_matrix = None
+        self.__current_character_matrix = None
+        if character_matrix is not None:
+            self.__original_character_matrix = character_matrix.copy()
+            self.__current_character_matrix = character_matrix.copy()
+        
+        self.missing_state_indicator = missing_state_indicator
```

Comment:
Yeah, that's another good idea @richardyz98 , it would allow us to easily parse data like that of the DREAM subchallenge 1 without having to re-map the states ourselves first.